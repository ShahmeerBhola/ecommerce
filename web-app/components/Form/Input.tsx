import React from 'react';

import Text from '../Text';

type Props = {
    name: string;
    type: string;
    value: string;
    placeholder?: string;
    disabled?: boolean;
    overrideContent?: () => JSX.Element;
    borderColor?: 'border-red-200' | 'border-gray-901';
    onChange?: Forms.InputOnChange;
    onBlur?: Forms.InputOnBlur;
    blankOptionsFallbackText?: string;
    options?: Forms.FieldOption[];
};

const Input: React.FunctionComponent<Props> = ({
    name,
    type,
    value,
    placeholder,
    disabled,
    overrideContent,
    borderColor = 'border-gray-901',
    onChange,
    onBlur,
    blankOptionsFallbackText,
    options,
}) => {
    const baseStyle = 'text-sm text-gray-902 placeholder-gray-902 p-3 focus:outline-none';
    const inputStyle = `${baseStyle} w-full h-50px bg-gray-50 border border-solid ${borderColor}`;
    const textAreaStyle = `${baseStyle} block w-full h-175px bg-gray-50 border border-solid ${borderColor}`;
    const checkboxRadioStyle = `${baseStyle} mr-5`;

    if (type === 'radio') {
        // Radio buttons seem to be a bug in Formik..
        // see here for Github Issue https://github.com/jaredpalmer/formik/issues/1024
        // and here for a solution https://codesandbox.io/s/88jqoy4qvl
        const radioOptions = options as Forms.FieldOption[];

        if (overrideContent) {
            return overrideContent();
        }

        if (radioOptions.length === 0) {
            return <Text>{blankOptionsFallbackText}</Text>;
        }

        return (
            <>
                {placeholder && <Text>{placeholder}</Text>}
                {radioOptions.map(({ value: radioValue, label: radioLabel }) => (
                    <div key={radioValue}>
                        <label htmlFor={`${name}_${radioValue}`}>
                            <input
                                id={`${name}_${radioValue}`}
                                name={name}
                                type="radio"
                                disabled={disabled}
                                value={radioValue}
                                defaultChecked={value === radioValue}
                                onChange={onChange}
                                onBlur={onBlur}
                                className={checkboxRadioStyle}
                            />
                            {radioLabel}
                        </label>
                    </div>
                ))}
            </>
        );
    } else if (type === 'checkbox') {
        // Formik checkboxes are really stubborn...
        // https://github.com/formium/formik/issues/602
        return (
            <>
                <input
                    name={name}
                    type={type}
                    value={value}
                    disabled={disabled}
                    defaultChecked={(value as any) as boolean}
                    className={checkboxRadioStyle}
                    placeholder={placeholder}
                    onChange={onChange}
                    onBlur={onBlur}
                />
                <label htmlFor={name}>{placeholder}</label>
            </>
        );
    } else if (type === 'textarea') {
        return (
            <textarea
                name={name}
                rows={4}
                value={value}
                placeholder={placeholder}
                className={textAreaStyle}
                onChange={onChange}
                onBlur={onBlur}
            />
        );
    }

    return (
        <input
            name={name}
            type={type}
            value={value}
            disabled={disabled}
            className={inputStyle}
            placeholder={placeholder}
            onChange={onChange}
            onBlur={onBlur}
        />
    );
};

export default Input;
