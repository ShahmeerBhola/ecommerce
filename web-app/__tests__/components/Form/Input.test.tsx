import React from 'react';
import renderer from 'react-test-renderer';
import Input from '../../../components/Form/Input';

describe('Input component', () => {
  const radioOptions = [
    {
      value: 'foo',
      label: 'Foo',
    },
    {
      value: 'bar',
      label: 'Bar',
    },
  ];

  test('it renders properly for radio type when value is blank', () => {
    const tree = renderer.create(<Input name="test_radio" type="radio" value="" options={radioOptions} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for radio type when value is non-blank', () => {
    const tree = renderer.create(<Input name="test_radio" type="radio" value="foo" options={radioOptions} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for radio type with placeholder', () => {
    const tree = renderer
      .create(
        <Input name="test_radio" placeholder="This is the placeholder" type="radio" value="" options={radioOptions} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for radio type when placeholder is blank string', () => {
    const tree = renderer
      .create(<Input name="test_radio" placeholder="" type="radio" value="" options={radioOptions} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for radio type when options is an empty array', () => {
    const tree = renderer.create(<Input name="test_radio" type="radio" value="" options={[]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for radio type when onChange and onBlur are provided', () => {
    const tree = renderer
      .create(
        <Input
          name="test_radio"
          type="radio"
          value=""
          options={radioOptions}
          onChange={(): void => {
            console.log('hello');
          }}
          onBlur={(): void => {
            console.log('hello');
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for radio type when disabled prop is false', () => {
    const tree = renderer
      .create(<Input name="test_radio" type="radio" disabled={false} value="" options={radioOptions} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for radio type when disabled prop is true', () => {
    const tree = renderer
      .create(<Input name="test_radio" type="radio" disabled value="" options={radioOptions} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for checkbox type', () => {
    const tree = renderer
      .create(<Input name="test_checkbox" type="checkbox" value="true" placeholder="Placeholder" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for checkbox type when disabled is false', () => {
    const tree = renderer
      .create(<Input name="test_checkbox" type="checkbox" value="false" disabled={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for checkbox type when disabled is true', () => {
    const tree = renderer.create(<Input name="test_checkbox" type="checkbox" value="false" disabled />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for textarea type', () => {
    const tree = renderer
      .create(
        <Input
          name="test_textarea"
          type="textarea"
          value="yoooooo"
          placeholder="This is the placeholder"
          onChange={(): void => {
            console.log('hello');
          }}
          onBlur={(): void => {
            console.log('hello');
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly for other types', () => {
    const tree = renderer
      .create(
        <Input
          name="test_text"
          type="text"
          value="yoooooo"
          disabled
          placeholder="This is the placeholder"
          onChange={(): void => {
            console.log('hello');
          }}
          onBlur={(): void => {
            console.log('hello');
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when borderColor prop is supplied', () => {
    const tree = renderer.create(<Input name="test_text" type="text" value="" borderColor="border-red-200" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
