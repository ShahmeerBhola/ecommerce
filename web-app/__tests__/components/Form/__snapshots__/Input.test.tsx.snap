// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component it renders properly for checkbox type 1`] = `
Array [
  <input
    className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
    name="test_checkbox"
    placeholder="Placeholder"
    type="checkbox"
    value="true"
  />,
  <label
    htmlFor="test_checkbox"
  >
    Placeholder
  </label>,
]
`;

exports[`Input component it renders properly for checkbox type when disabled is false 1`] = `
Array [
  <input
    className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
    disabled={false}
    name="test_checkbox"
    type="checkbox"
    value="false"
  />,
  <label
    htmlFor="test_checkbox"
  />,
]
`;

exports[`Input component it renders properly for checkbox type when disabled is true 1`] = `
Array [
  <input
    className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
    disabled={true}
    name="test_checkbox"
    type="checkbox"
    value="false"
  />,
  <label
    htmlFor="test_checkbox"
  />,
]
`;

exports[`Input component it renders properly for other types 1`] = `
<input
  className="text-sm text-gray-200 p-3 focus:outline-none w-full h-50px bg-gray-50 border border-solid border-gray-100"
  disabled={true}
  name="test_text"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="This is the placeholder"
  type="text"
  value="yoooooo"
/>
`;

exports[`Input component it renders properly for radio type when disabled prop is false 1`] = `
Array [
  <div>
    <label
      htmlFor="test_radio_foo"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        disabled={false}
        id="test_radio_foo"
        name="test_radio"
        type="radio"
        value="foo"
      />
      Foo
    </label>
  </div>,
  <div>
    <label
      htmlFor="test_radio_bar"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        disabled={false}
        id="test_radio_bar"
        name="test_radio"
        type="radio"
        value="bar"
      />
      Bar
    </label>
  </div>,
]
`;

exports[`Input component it renders properly for radio type when disabled prop is true 1`] = `
Array [
  <div>
    <label
      htmlFor="test_radio_foo"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        disabled={true}
        id="test_radio_foo"
        name="test_radio"
        type="radio"
        value="foo"
      />
      Foo
    </label>
  </div>,
  <div>
    <label
      htmlFor="test_radio_bar"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        disabled={true}
        id="test_radio_bar"
        name="test_radio"
        type="radio"
        value="bar"
      />
      Bar
    </label>
  </div>,
]
`;

exports[`Input component it renders properly for radio type when onChange and onBlur are provided 1`] = `
Array [
  <div>
    <label
      htmlFor="test_radio_foo"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_foo"
        name="test_radio"
        onBlur={[Function]}
        onChange={[Function]}
        type="radio"
        value="foo"
      />
      Foo
    </label>
  </div>,
  <div>
    <label
      htmlFor="test_radio_bar"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_bar"
        name="test_radio"
        onBlur={[Function]}
        onChange={[Function]}
        type="radio"
        value="bar"
      />
      Bar
    </label>
  </div>,
]
`;

exports[`Input component it renders properly for radio type when options is an empty array 1`] = `
<p
  className="font-secondary font-normal text-gray-100 text-base"
/>
`;

exports[`Input component it renders properly for radio type when placeholder is blank string 1`] = `
Array [
  "",
  <div>
    <label
      htmlFor="test_radio_foo"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_foo"
        name="test_radio"
        type="radio"
        value="foo"
      />
      Foo
    </label>
  </div>,
  <div>
    <label
      htmlFor="test_radio_bar"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_bar"
        name="test_radio"
        type="radio"
        value="bar"
      />
      Bar
    </label>
  </div>,
]
`;

exports[`Input component it renders properly for radio type when value is blank 1`] = `
Array [
  <div>
    <label
      htmlFor="test_radio_foo"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_foo"
        name="test_radio"
        type="radio"
        value="foo"
      />
      Foo
    </label>
  </div>,
  <div>
    <label
      htmlFor="test_radio_bar"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_bar"
        name="test_radio"
        type="radio"
        value="bar"
      />
      Bar
    </label>
  </div>,
]
`;

exports[`Input component it renders properly for radio type when value is non-blank 1`] = `
Array [
  <div>
    <label
      htmlFor="test_radio_foo"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={true}
        id="test_radio_foo"
        name="test_radio"
        type="radio"
        value="foo"
      />
      Foo
    </label>
  </div>,
  <div>
    <label
      htmlFor="test_radio_bar"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_bar"
        name="test_radio"
        type="radio"
        value="bar"
      />
      Bar
    </label>
  </div>,
]
`;

exports[`Input component it renders properly for radio type with placeholder 1`] = `
Array [
  <p
    className="font-secondary font-normal text-gray-100 text-base"
  >
    This is the placeholder
  </p>,
  <div>
    <label
      htmlFor="test_radio_foo"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_foo"
        name="test_radio"
        type="radio"
        value="foo"
      />
      Foo
    </label>
  </div>,
  <div>
    <label
      htmlFor="test_radio_bar"
    >
      <input
        className="text-sm text-gray-200 p-3 focus:outline-none mr-5"
        defaultChecked={false}
        id="test_radio_bar"
        name="test_radio"
        type="radio"
        value="bar"
      />
      Bar
    </label>
  </div>,
]
`;

exports[`Input component it renders properly for textarea type 1`] = `
<textarea
  className="text-sm text-gray-200 p-3 focus:outline-none w-full h-50px bg-gray-50 border border-solid border-gray-100"
  name="test_textarea"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="This is the placeholder"
  rows={4}
  value="yoooooo"
/>
`;

exports[`Input component it renders properly when borderColor prop is supplied 1`] = `
<input
  className="text-sm text-gray-200 p-3 focus:outline-none w-full h-50px bg-gray-50 border border-solid border-red-200"
  name="test_text"
  type="text"
  value=""
/>
`;
